### ISSUE TITLE
Type projection is unsound

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#1050]
The following program compiles and throws a ClassCastException when executed. One might be tempted to try to rule out the obviously bad intersection `T & U` but we know from our work on DOT that there are ways to engineer bad bounds that cannot be detected locally. So the culprit is the projection
`X#A`. 

    object Test {

      trait C { type A }

      type T = C { type A >: Any }
      type U = C { type A <: Nothing }
      type X = T & U

      def main(args: Array[String]) = {
        val y: X#A = 1
        val z: String = y
      }
    }

**Comment by smarter** [comment]
this looks very similar to https://github.com/lampepfl/dotty/issues/50

**Comment by adriaanm** [comment]
it kind of looks like an existential whose skolem is used outside of the existential scope -- when you open `X#A` to admit `1`, there's this unknown type that we assume for it, and then we carry that over and make another assumption on it for `z = y`.

**Comment by smarter** [comment]
I don't want to insist too much on this since I haven't done any more work to convince anyone but @sstucki that this makes sense, but my [type projection proposal](https://gist.github.com/smarter/831a8b6f8641bd857bfa) side-steps this: `X#A` is always considered to be a subtype and a supertype of its upper-bound, so you can never use it to do unsound casts.

**Comment by odersky** [comment]
@smarter Yes, but we plan to address #50 tightening the isVolatile check. We have not contemplated volatile checking for prefixes of projections. One could try that, but I fear it would throw out many interesting use cases of projections.

**Comment by odersky** [comment]
@smarter Here is the same example with aliases:

    object Test {

      trait C { type A }

      type T = C { type A = Any }
      type U = C { type A = Nothing }
      type X = T & U

      def main(args: Array[String]) = {
        val y: X#A = 1
        val z: String = y
      }
    }

Same ClassCastException. The way I interpret your proposal it means I map the first program to the second one, no?

**Comment by adriaanm** [comment]
Can you rewrite this using existential types and still get away with it?

```
val y: T forSome { type T :> Int} = 1
val z: String = y: T forSome { type T <: String }
```

**Comment by odersky** [comment]
We don't have existentials anymore.

**Comment by smarter** [comment]
@odersky :
> Same ClassCastException

This shouldn't happen and looks like a separate bug. I expect `A` in `X = T & U` to have type `>: (Any | Nothing) <: (Any & Nothing)`, which means that `val x: X#A = 1` should not compile because `Int` is not a subtype of `Any & Nothing`.

**Comment by adriaanm** [comment]
Sure, it was a thought experiment to see how they relate.

**Comment by adriaanm** [comment]
How about saying that `val z: String = y` only type checks if all of `y`'s bounds (lower and upper) are a subtype of `String`?

**Comment by odersky** [comment]
@smarter No, in the example Int needs to be a subtype of the lower bound, `Any | Nothing`, which it is, obviously.

**Comment by smarter** [comment]
It looks like we have three possible set of rules, which look a bit like this (very approximately):

- current:
```
------------------------------
G |- {L: X..Y, ...}#L <: Y

------------------------------
G |- X <: {L: X..Y, ...}#L
```

-  @adriaanm's proposal above, which I find interesting:

```
------------------------------
G |- {L: X..Y, ...}#L <: Y

G |- T <: X     G |- T <: Y
------------------------------
G |- T <: {L: X..Y, ...}#L
```

- my proposal: see https://gist.github.com/smarter/831a8b6f8641bd857bfa

**Comment by smarter** [comment]
> @smarter No, in the example Int needs to be a subtype of the lower bound, Any | Nothing, which it is, obviously. 

Right, that's different from what would happen in my proposal where `Int` needs to be a subtype of the upper bound.

**Comment by adriaanm** [comment]
(Note that I withdrew my proposal ;-) `T <: X` doesn't make sense in general, since `X` defaults to `Nothing`. It could work out better to flip it: `X <: T`.)

**Comment by smarter** [comment]
>  T <: X doesn't make sense in general, since X defaults to Nothing

How is that different from what we're doing now? In both scalac and dotty, if you write `trait A { type X <: Any}`, then the only subtype of `A#X` is Nothing.

**Comment by odersky** [comment]
Closed by way of #1051
